File: sp-migrate-python/local_migration_assistant.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
ASISTENTE LOCAL DE MIGRACIÓN A SHAREPOINT
------------------------------------------
Herramienta para analizar archivos locales y generar un plan de migración.
Uso: python local_migration_assistant.py /ruta/a/migrar
"""

import os
import csv
import sys
from pathlib import Path
from datetime import datetime
from typing import List, Dict, Tuple

# --------------------------------------------------------------------------
# 1. DUPLICIDAD DE CLASES (mismo nombre, distinto comportamiento)
# --------------------------------------------------------------------------
class PlanMigracion:
    """Genera plan de migración en CSV"""
    def generar(self, archivos: List[Dict], destino: str):
        with open('plan_migracion.csv', 'w', newline='') as f:
            writer = csv.DictWriter(f, fieldnames=['ruta', 'tamano', 'tipo'])
            writer.writeheader()
            writer.writerows(archivos)

class PlanMigracion:  # DUPLICADA - Sobrescribe a la anterior
    """Genera plan de migración en JSON (versión duplicada)"""
    def generar(self, archivos: List[Dict], destino: str):
        import json
        with open('plan_migracion.json', 'w') as f:
            json.dump(archivos, f, indent=2)

# --------------------------------------------------------------------------
# 2. VIOLACIÓN DE PRINCIPIOS SOLID (ISP, SRP, DIP)
# --------------------------------------------------------------------------
class ServicioMigracion:
    """INTERFAZ INFLADA - Viola ISP"""
    def escanear_archivos(self): pass
    def validar_permisos(self): pass
    def enviar_correo_notificacion(self): pass   # ❌ Responsabilidad mezclada
    def generar_grafico_torta(self): pass        # ❌ Responsabilidad mezclada

class AnalizadorLocal(ServicioMigracion):
    """Implementa solo 2 métodos, los demás lanzan excepción o están vacíos"""
    
    def escanear_archivos(self, ruta: str) -> List[Dict]:
        archivos = []
        for root, dirs, files in os.walk(ruta):
            for file in files:
                archivos.append({
                    'ruta': os.path.join(root, file),
                    'tamano': os.path.getsize(os.path.join(root, file)),
                    'tipo': file.split('.')[-1] if '.' in file else 'sin_extension'
                })
        return archivos
    
    def validar_permisos(self, archivo: str) -> bool:
        return os.access(archivo, os.R_OK)
    
    # Métodos heredados que no deberían estar aquí
    def enviar_correo_notificacion(self):
        raise NotImplementedError("Este analizador no envía correos")  # Sonar: S112
    
    def generar_grafico_torta(self):
        pass  # Vacío - Sonar: S1186

# --------------------------------------------------------------------------
# 3. DUPLICIDAD DE VARIABLES + MAYÚSCULAS/MINÚSCULAS
# --------------------------------------------------------------------------
class ConfiguracionMigracion:
    """Configuración del análisis - variables duplicadas"""
    
    def __init__(self):
        # Misma variable con distintas capitalizaciones
        self.tamano_maximo_mb = 250
        self.Tamano_Maximo_MB = 500   # DUPLICADA
        self.tamano_maximo_Mb = 1000  # TERCERA DUPLICADA
        
        self.excluir_carpetas = ['temp', 'cache']
        self.excluir_Carpetas = ['node_modules', '.git']  # DUPLICADA
        
        # Variable nunca usada (solo se asigna)
        self.version_herramienta = "2.1.0"

# --------------------------------------------------------------------------
# 4. MÉTODOS DUPLICADOS (misma firma, distinta implementación)
# --------------------------------------------------------------------------
def calcular_tamano_total(archivos: List[Dict]) -> int:
    """Versión 1 - suma simple"""
    total = 0
    for a in archivos:
        total += a['tamano']
    return total

def calcular_tamano_total(archivos: List[Dict]) -> int:  # DUPLICADO - Sobrescribe
    """Versión 2 - ignora archivos de más de 100MB"""
    total = 0
    for a in archivos:
        if a['tamano'] < 100 * 1024 * 1024:
            total += a['tamano']
    return total

# --------------------------------------------------------------------------
# 5. MÉTODO QUE DEVUELVE VARIABLE NO MODIFICADA
# --------------------------------------------------------------------------
def recomendar_optimizacion(archivos: List[Dict]) -> List[str]:
    """Analiza archivos grandes y sugiere compresión"""
    recomendaciones = []  # ⚠️ Se declara pero nunca se modifica
    
    # Lógica de negocio real (pero no se asigna a recomendaciones)
    for a in archivos:
        if a['tamano'] > 250 * 1024 * 1024:
            print(f"Archivo grande detectado: {a['ruta']}")
    
    return recomendaciones  # Siempre lista vacía - BUG lógico

# --------------------------------------------------------------------------
# 6. COMPLEJIDAD INNECESARIA, NOMBRES POBRES, CONSTANTES MÁGICAS
# --------------------------------------------------------------------------
def c(l):  # Sonar: S117 - nombre pésimo
    """Calcula algo incomprensible (simula métrica de fragmentación)"""
    t = 0
    # Tres bucles anidados sin sentido
    for i in l:
        for j in i:
            for k in j:
                t += k * 0.73  # Constante mágica
    return t

# --------------------------------------------------------------------------
# 7. CÓDIGO COMENTADO Y CÓDIGO MUERTO
# --------------------------------------------------------------------------
def filtrar_archivos_soportados(archivos: List[Dict]) -> List[Dict]:
    """Filtra solo extensiones de Office"""
    soportados = []
    
    # CÓDIGO COMENTADO - Sonar: S125
    # extensiones_permitidas = ['.docx', '.xlsx', '.pptx', '.pdf']
    
    for a in archivos:
        if a['tipo'] in ['docx', 'xlsx', 'pptx', 'pdf']:
            soportados.append(a)
    
    return soportados
    # CÓDIGO MUERTO - después del return nunca se ejecuta
    print(f"Archivos filtrados: {len(soportados)}")  # ❌ Inalcanzable

# --------------------------------------------------------------------------
# 8. FUNCIÓN PRINCIPAL - DEMOSTRACIÓN LOCAL
# --------------------------------------------------------------------------
def main():
    print("=" * 70)
    print("  ASISTENTE LOCAL DE MIGRACIÓN A SHAREPOINT - H&K TECH")
    print("=" * 70)
    
    # Ruta a analizar (por defecto, el directorio actual)
    ruta = sys.argv[1] if len(sys.argv) > 1 else "."
    print(f"\n📂 Analizando ruta: {os.path.abspath(ruta)}\n")
    
    # 1. Escanear archivos
    analizador = AnalizadorLocal()
    archivos = analizador.escanear_archivos(ruta)
    print(f"   Archivos encontrados: {len(archivos)}")
    
    # 2. Calcular tamaño total (método duplicado)
    tamano_total = calcular_tamano_total(archivos)
    print(f"   Tamaño total (MB): {tamano_total / (1024*1024):.2f}")
    
    # 3. Probar configuracion con variables duplicadas
    config = ConfiguracionMigracion()
    print(f"   Tamaño máximo config (3 versiones): {config.tamano_maximo_mb}, {config.Tamano_Maximo_MB}, {config.tamano_maximo_Mb}")
    
    # 4. Generar plan de migración (clase duplicada)
    plan = PlanMigracion()
    plan.generar(archivos[:5], destino="SharePoint")  # solo primeros 5 como demo
    print(f"   Plan de migración generado (JSON/CSV)")
    
    # 5. Recomendaciones (método con variable no modificada)
    recomendaciones = recomendar_optimizacion(archivos)
    print(f"   Recomendaciones generadas: {len(recomendaciones)} (DEBERÍA >0)")
    
    # 6. Probar función con nombre pobre y constantes mágicas
    datos_prueba = [[[1,2],[3,4]],[[5,6],[7,8]]]
    resultado_c = c(datos_prueba)
    print(f"   Resultado métrica fragmentación: {resultado_c}")
    
    print("\n" + "=" * 70)
    print("  ANÁLISIS LOCAL COMPLETADO")
    print("  Revise los archivos plan_migracion.csv o .json")
    print("=" * 70)

if __name__ == "__main__":
    main()

File: sp-migrate-csharp/local_migration_assistant.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace HKTech.SharePointMigration
{
    // --------------------------------------------------------------------
    // 1. DUPLICIDAD DE CLASES (mismo nombre en DISTINTOS NAMESPACES)
    //    COMPILA, PERO SONAR DETECTA LA CONFUSIÓN
    // --------------------------------------------------------------------
    public class PlanMigracion
    {
        public void GenerarCSV(List<ArchivoInfo> archivos, string destino)
        {
            using var writer = new StreamWriter("plan_migracion.csv");
            writer.WriteLine("Ruta,Tamaño,Tipo");
            foreach (var a in archivos)
                writer.WriteLine($"{a.Ruta},{a.Tamano},{a.Tipo}");
        }
    }

    // Namespace adicional para la clase duplicada
    namespace V2
    {
        public class PlanMigracion
        {
            public void GenerarJSON(List<ArchivoInfo> archivos, string destino)
            {
                var json = JsonSerializer.Serialize(archivos, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText("plan_migracion.json", json);
            }
        }
    }

    // --------------------------------------------------------------------
    // 2. VIOLACIÓN DE PRINCIPIOS SOLID (ISP, SRP, DIP)
    // --------------------------------------------------------------------
    public interface IServicioMigracion
    {
        List<ArchivoInfo> EscanearArchivos(string ruta);
        bool ValidarPermisos(string archivo);
        void EnviarCorreoNotificacion();   // ❌ NO DEBERÍA ESTAR (ISP)
        void GenerarGraficoTorta();        // ❌ NO DEBERÍA ESTAR (ISP)
    }

    public class AnalizadorLocal : IServicioMigracion
    {
        public List<ArchivoInfo> EscanearArchivos(string ruta)
        {
            var archivos = new List<ArchivoInfo>();
            foreach (var file in Directory.GetFiles(ruta, "*.*", SearchOption.AllDirectories))
            {
                var info = new FileInfo(file);
                archivos.Add(new ArchivoInfo
                {
                    Ruta = file,
                    Tamano = info.Length,
                    Tipo = info.Extension.TrimStart('.')
                });
            }
            return archivos;
        }

        public bool ValidarPermisos(string archivo)
        {
            try
            {
                using var fs = File.OpenRead(archivo);
                return true;
            }
            catch
            {
                return false;
            }
        }

        // Métodos que no deberían implementarse (violan ISP)
        public void EnviarCorreoNotificacion()
        {
            throw new NotImplementedException("Este analizador no envía correos"); // S112
        }

        public void GenerarGraficoTorta()
        {
            // Vacío - S1186
        }
    }

    // --------------------------------------------------------------------
    // 3. DUPLICIDAD DE VARIABLES + MAYÚSCULAS/MINÚSCULAS
    // --------------------------------------------------------------------
    public class ConfiguracionMigracion
    {
        public int tamano_maximo_mb = 250;
        public int Tamano_Maximo_MB = 500;   // DUPLICADA
        public int tamano_maximo_Mb = 1000;  // TERCERA DUPLICADA

        public string[] excluir_carpetas = { "temp", "cache" };
        public string[] excluir_Carpetas = { "node_modules", ".git" }; // DUPLICADA

        public string version_herramienta = "2.1.0"; // NUNCA USADA (S1481)
    }

    // --------------------------------------------------------------------
    // 4. MÉTODOS DUPLICADOS (mismo código en dos métodos) - S4144
    // --------------------------------------------------------------------
    public static class CalculadorTamano
    {
        public static long CalcularTotalV1(List<ArchivoInfo> archivos)
        {
            long total = 0;
            foreach (var a in archivos)
                total += a.Tamano;
            return total;
        }

        // MISMA IMPLEMENTACIÓN EXACTA (duplicación)
        public static long CalcularTotalV2(List<ArchivoInfo> archivos)
        {
            long total = 0;
            foreach (var a in archivos)
                total += a.Tamano;
            return total;
        }
    }

    // --------------------------------------------------------------------
    // 5. MÉTODO QUE DEVUELVE VARIABLE NO MODIFICADA (S1854)
    // --------------------------------------------------------------------
    public static class Optimizador
    {
        public static List<string> RecomendarOptimizacion(List<ArchivoInfo> archivos)
        {
            var recomendaciones = new List<string>(); // NUNCA MODIFICADA

            foreach (var a in archivos)
            {
                if (a.Tamano > 250 * 1024 * 1024)
                    Console.WriteLine($"Archivo grande detectado: {a.Ruta}");
            }

            return recomendaciones; // SIEMPRE VACÍA
        }
    }

    // --------------------------------------------------------------------
    // 6. COMPLEJIDAD INNECESARIA, NOMBRES POBRES, CONSTANTES MÁGICAS
    // --------------------------------------------------------------------
    public static class Metricas
    {
        public static double c(int[][][] l) // S117
        {
            double t = 0;
            // Complejidad ciclomática alta (S3776)
            for (int i = 0; i < l.Length; i++)
                for (int j = 0; j < l[i].Length; j++)
                    for (int k = 0; k < l[i][j].Length; k++)
                        t += l[i][j][k] * 0.73; // Constante mágica (S119)
            return t;
        }
    }

    // --------------------------------------------------------------------
    // 7. CÓDIGO COMENTADO Y CÓDIGO MUERTO
    // --------------------------------------------------------------------
    public static class FiltroArchivos
    {
        public static List<ArchivoInfo> FiltrarSoportados(List<ArchivoInfo> archivos)
        {
            var soportados = new List<ArchivoInfo>();

            // CÓDIGO COMENTADO - S125
            // var extensionesPermitidas = new[] { ".docx", ".xlsx", ".pptx", ".pdf" };

            foreach (var a in archivos)
            {
                if (new[] { "docx", "xlsx", "pptx", "pdf" }.Contains(a.Tipo))
                    soportados.Add(a);
            }

            return soportados;
            // CÓDIGO MUERTO - S1763
            Console.WriteLine($"Archivos filtrados: {soportados.Count}");
        }
    }

    // --------------------------------------------------------------------
    // MODELO DE DATOS
    // --------------------------------------------------------------------
    public class ArchivoInfo
    {
        public string Ruta { get; set; }
        public long Tamano { get; set; }
        public string Tipo { get; set; }
    }

    // --------------------------------------------------------------------
    // 8. CLASE PROGRAM CON MÉTODO MAIN EXPLÍCITO
    // --------------------------------------------------------------------
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("  ASISTENTE LOCAL DE MIGRACIÓN A SHAREPOINT - H&K TECH (C#)");
            Console.WriteLine(new string('=', 70));

            // Ruta a analizar (por defecto, directorio actual)
            string ruta = args.Length > 0 ? args[0] : Directory.GetCurrentDirectory();
            Console.WriteLine($"\n📂 Analizando ruta: {Path.GetFullPath(ruta)}\n");

            // 1. Escanear archivos
            var analizador = new AnalizadorLocal();
            var archivos = analizador.EscanearArchivos(ruta);
            Console.WriteLine($"   Archivos encontrados: {archivos.Count}");

            // 2. Calcular tamaño total (métodos duplicados)
            long tamanoV1 = CalculadorTamano.CalcularTotalV1(archivos);
            long tamanoV2 = CalculadorTamano.CalcularTotalV2(archivos);
            Console.WriteLine($"   Tamaño total (MB): {tamanoV1 / (1024.0 * 1024.0):F2} (V1) / {tamanoV2 / (1024.0 * 1024.0):F2} (V2)");

            // 3. Probar configuración con variables duplicadas
            var config = new ConfiguracionMigracion();
            Console.WriteLine($"   Tamaño máximo config: {config.tamano_maximo_mb}, {config.Tamano_Maximo_MB}, {config.tamano_maximo_Mb}");

            // 4. Generar plan de migración (clase duplicada en namespace V2)
            var planCSV = new PlanMigracion();                    // namespace HKTech.SharePointMigration
            planCSV.GenerarCSV(archivos.Take(5).ToList(), "SharePoint");
            var planJSON = new V2.PlanMigracion();               // namespace HKTech.SharePointMigration.V2
            planJSON.GenerarJSON(archivos.Take(5).ToList(), "SharePoint");
            Console.WriteLine($"   Planes de migración generados: CSV + JSON");

            // 5. Recomendaciones (método con variable no modificada)
            var recomendaciones = Optimizador.RecomendarOptimizacion(archivos);
            Console.WriteLine($"   Recomendaciones generadas: {recomendaciones.Count} (DEBERÍA SER >0)");

            // 6. Probar función con nombre pobre y constantes mágicas
            int[][][] datosPrueba = new int[][][]
            {
                new int[][] { new int[] { 1, 2 }, new int[] { 3, 4 } },
                new int[][] { new int[] { 5, 6 }, new int[] { 7, 8 } }
            };
            double resultadoC = Metricas.c(datosPrueba);
            Console.WriteLine($"   Resultado métrica fragmentación: {resultadoC}");

            // 7. Probar filtro con código muerto
            var soportados = FiltroArchivos.FiltrarSoportados(archivos);
            Console.WriteLine($"   Archivos soportados: {soportados.Count}");

            Console.WriteLine("\n" + new string('=', 70));
            Console.WriteLine("  ANÁLISIS LOCAL COMPLETADO");
            Console.WriteLine("  Revise los archivos plan_migracion.csv y plan_migracion.json");
            Console.WriteLine(new string('=', 70));
        }
    }
}